internal/newstate/graph_test.go

```diff
@@ -0,0 +1,615 @@
+				}),
+			},
+			gwNsName: types.NamespacedName{Namespace: "test", Name: "gateway"},
+			listeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			expectedRoutes: map[types.NamespacedName]*route{},
+			expectedListeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			msg: "HTTPRoute without good parent refs",
+		},
+		{
+			httpRoutes: map[types.NamespacedName]*v1alpha2.HTTPRoute{
+				{Namespace: "test", Name: "hr-1"}: hrNonExistingSectionName,
+			},
+			gwNsName: types.NamespacedName{Namespace: "test", Name: "gateway"},
+			listeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			expectedRoutes: map[types.NamespacedName]*route{
+				{Namespace: "test", Name: "hr-1"}: {
+					Source:               hrNonExistingSectionName,
+					ValidSectionNameRefs: map[string]struct{}{},
+					InvalidSectionNameRefs: map[string]struct{}{
+						"listener-80-2": {},
+					},
+				},
+			},
+			expectedListeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			msg: "HTTPRoute with non-existing section name",
+		},
+		{
+			httpRoutes: map[types.NamespacedName]*v1alpha2.HTTPRoute{
+				{Namespace: "test", Name: "hr-1"}: hrFoo,
+			},
+			gwNsName: types.NamespacedName{Namespace: "test", Name: "gateway"},
+			listeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			expectedRoutes: map[types.NamespacedName]*route{
+				{Namespace: "test", Name: "hr-1"}: {
+					Source: hrFoo,
+					ValidSectionNameRefs: map[string]struct{}{
+						"listener-80-1": {},
+					},
+					InvalidSectionNameRefs: map[string]struct{}{},
+				},
+			},
+			expectedListeners: map[string]*listener{
+				"listener-80-1": createModifiedListener(func(l *listener) {
+					l.Routes = map[types.NamespacedName]*route{
+						{Namespace: "test", Name: "hr-1"}: {
+							Source: hrFoo,
+							ValidSectionNameRefs: map[string]struct{}{
+								"listener-80-1": {},
+							},
+							InvalidSectionNameRefs: map[string]struct{}{},
+						},
+					}
+					l.AcceptedHostnames = map[string]struct{}{
+						"foo.example.com": {},
+					}
+				}),
+			},
+			msg: "HTTPRoute with one accepted hostname",
+		},
+		{
+			httpRoutes: map[types.NamespacedName]*v1alpha2.HTTPRoute{
+				{Namespace: "test", Name: "hr-1"}: hrBar,
+			},
+			gwNsName: types.NamespacedName{Namespace: "test", Name: "gateway"},
+			listeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			expectedRoutes: map[types.NamespacedName]*route{
+				{Namespace: "test", Name: "hr-1"}: {
+					Source:               hrBar,
+					ValidSectionNameRefs: map[string]struct{}{},
+					InvalidSectionNameRefs: map[string]struct{}{
+						"listener-80-1": {},
+					},
+				},
+			},
+			expectedListeners: map[string]*listener{
+				"listener-80-1": createListener(),
+			},
+			msg: "HTTPRoute with zero accepted hostnames",
+		},
+	}
```

Comment:
I refactored buildRoutesAndBindToListeners so that it only deals with one httpRoute - there is no need for that test cases.
I moved the loop into buildGraph and added a second HTTPRoute in TestBuildGraph.
Let me know.